File: 03-stream-processing/solution-proposal/scala/src/main/scala/com/becomedataengineer/exercise2/ReprocessingJob.scala
package com.becomedataengineer.exercise2

import com.becomedataengineer.exercise2.SinkConfiguration.SinkOptions
import org.apache.spark.sql.SparkSession

import java.time.LocalDate
import java.time.format.DateTimeFormatter
import scala.collection.mutable

object ReprocessingJob {

  def main(args: Array[String]): Unit = {
    val firstDate = LocalDate.parse(args(0))
    val lastDate = LocalDate.parse(args(1))
    var partitionDate = firstDate
    val partitionDates = new mutable.ListBuffer[String]()
    while (partitionDate.isBefore(lastDate.plusDays(1))) {
      partitionDates.append(partitionDate.format(DateTimeFormatter.ISO_DATE))
      partitionDate = partitionDate.plusDays(1)
    }
    val sparkSession = SparkSession.builder().master("local[*]")
      .getOrCreate()

    val inputDataFrames = partitionDates.map(eventTime => {
      sparkSession.read.schema("key STRING, value STRING").json(s"${FilesOutputDir}/event_time=${eventTime}")
    })

    val allDataToReprocess = inputDataFrames
      .reduce((partitionsSoFar, newPartition) => partitionsSoFar.union(newPartition))

    val recordsToWrite = StreamingJobProcessingLogic.processInputData(allDataToReprocess)

    recordsToWrite.write.format("kafka").options(SinkOptions).save()
  }

}

File: 03-stream-processing/solution-proposal/python/exercise2/reprocessing_job.py
import functools
import sys
from datetime import datetime, timedelta

from pyspark.sql import SparkSession

from exercise2.configuration import get_files_output_dir, get_sink_configuration
from exercise2.streaming_job_processing_logic import process_input_data


def reprocess_data_between_dates(start_date: datetime, end_date: datetime):
    spark = SparkSession.builder.master("local[*]") \
        .config('spark.jars.packages', 'org.apache.spark:spark-sql-kafka-0-10_2.12:3.4.0') \
        .getOrCreate()

    partition_dataframes_to_reprocess = []
    current_date = start_date
    while current_date <= end_date:
        event_time = str(current_date.strftime('%Y-%m-%d'))
        partition_dataframes_to_reprocess.append(
            spark.read.schema("key STRING, value STRING").json(f"{get_files_output_dir()}/event_time={event_time}")
        )
        current_date += timedelta(days=1)

    final_dataframe_to_reprocess = functools.reduce(
        lambda all_partitions, partition_df: all_partitions.union(partition_df),
        partition_dataframes_to_reprocess)

    records_to_write = process_input_data(final_dataframe_to_reprocess)

    records_to_write.write.format("kafka") \
        .options(**get_sink_configuration()) \
        .save()


if __name__ == '__main__':
    start_date = datetime.strptime(sys.argv[1], '%Y-%m-%d')
    end_date = datetime.strptime(sys.argv[2], '%Y-%m-%d')
    reprocess_data_between_dates(start_date, end_date)
